src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/Upsert/ConditionalUpsertResourceHandlerTests.cs

```diff
@@ -0,0 +1,211 @@
+    }
+
+    [Fact]
+    public async Task GivenAConditionalUpsertResourceHandler_WhenUserHasOnlyReadPermission_ThenUnauthorizedExceptionIsThrown()
+    {
+        // Arrange
+        _authService
+            .CheckAccess(DataActions.Read | DataActions.Write | DataActions.Search | DataActions.Update, CancellationToken.None)
+            .Returns(DataActions.Read);
+
+        var patient = Samples.GetDefaultPatient();
+        var conditionalParameters = new List<Tuple<string, string>> { new("name", "John") };
+        var request = new ConditionalUpsertResourceRequest(patient, conditionalParameters, null);
+
+        // Act & Assert
+        await Assert.ThrowsAsync<UnauthorizedFhirActionException>(() => _conditionalUpsertHandler.Handle(request, CancellationToken.None));
+    }
+
+    [Fact]
+    public async Task GivenAConditionalUpsertResourceHandler_WhenUserLacksAllPermissions_ThenUnauthorizedExceptionIsThrown()
+    {
+        // Arrange
+        _authService
+            .CheckAccess(DataActions.Read | DataActions.Write | DataActions.Search | DataActions.Update, CancellationToken.None)
+            .Returns(DataActions.None);
+
+        var patient = Samples.GetDefaultPatient();
+        var conditionalParameters = new List<Tuple<string, string>> { new("name", "John") };
+        var request = new ConditionalUpsertResourceRequest(patient, conditionalParameters, null);
+
+        // Act & Assert
+        await Assert.ThrowsAsync<UnauthorizedFhirActionException>(() => _conditionalUpsertHandler.Handle(request, CancellationToken.None));
+    }
+
+    [Fact]
+    public async Task GivenAConditionalUpsertResourceHandler_WhenNoMatchFoundAndResourceHasNoId_ThenCreateIsExecuted()
+    {
+        // Arrange - Setup for no matches to test create path
+        var searchResults = SearchResult.Empty();
+        searchResults.Results = new List<SearchResultEntry>(); // No matches
+        
+        _searchService.ConditionalSearchAsync(
+            Arg.Any<string>(),
+            Arg.Any<IReadOnlyList<Tuple<string, string>>>(),
+            Arg.Any<CancellationToken>(),
+            Arg.Any<ILogger>())
+            .Returns(searchResults);
+
+        _authService
+            .CheckAccess(DataActions.Read | DataActions.Write | DataActions.Search | DataActions.Update, CancellationToken.None)
+            .Returns(DataActions.Search | DataActions.Update);
+
+        var patient = Samples.GetDefaultPatient();
+        patient.Id = null; // No ID to trigger create path
+        var conditionalParameters = new List<Tuple<string, string>> { new("name", "John") };
+        var request = new ConditionalUpsertResourceRequest(patient, conditionalParameters, null);
+
+        // Act
+        await _conditionalUpsertHandler.Handle(request, CancellationToken.None);
+
+        // Assert - Should call create since no ID and no matches
+        await _mediator
```

Comment:
It will fail once it reach the CreateHandler since it has only Search | Update right?